### PR TITLE
Use freezed to modeling

### PR DESCRIPTION
**Pull request by go-finn-nguyen**
## Refactored model
 - [x] MHandle
 - [x] MResult
 - [x] MPagination
 - [x] MSocialUser
 - [x] MUser

### CODE REVIEW

File: lib/src/network/model/common/pagination/pagination.dart [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+abstract class MPagination<T> with _$MPagination<T> {
+  const MPagination._();
+  const factory MPagination({
+    @Default(15) int pageCount,
+    @Default(MHandle.initial()) MHandle<void> status,
+    @Default(true) bool hasMore,
+    @Default([]) List<T> items,
+  }) = _MPagination<T>;
+
+  bool get canNext => status.maybeWhen(
```
Comment (jimmy-to-goldenowl) [inline lib/src/network/model/common/pagination/pagination.dart]: bool get canNext {
return hasMore && status is initial or complete
}

File: lib/src/network/model/common/handle/handle.dart [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import 'package:myapp/src/network/model/common/result/result.dart';
+
+part 'handle.freezed.dart';
+
+@freezed
+class MHandle<T> with _$MHandle<T> {
+  const MHandle._();
+  const factory MHandle.initial() = _Initial;
+  const factory MHandle.loading() = _Loading;
+  const factory MHandle.complete(MResult<T> result) = _Complete<T>;
```
Comment (jimmy-to-goldenowl) [inline lib/src/network/model/common/handle/handle.dart:13]: how about error state?

File: lib/widgets/state/state_widget.dart [hunk 3]
```diff
@@ -1,21 +1,21 @@
-import 'package:flutter/material.dart';
-import 'package:myapp/src/network/model/common/handle.dart';
-import 'package:myapp/widgets/state/state_error_widget.dart';
-import 'package:myapp/widgets/state/state_loading_widget.dart';
+// import 'package:flutter/material.dart';
+// import 'package:myapp/src/network/model/common/handle.dart';
+// import 'package:myapp/widgets/state/state_error_widget.dart';
+// import 'package:myapp/widgets/state/state_loading_widget.dart';
 
-class XStateWidget extends StatelessWidget {
```
Comment (jimmy-to-goldenowl) [inline lib/widgets/state/state_widget.dart]: Could you help to update this file too?

**Review by jimmy-to-goldenowl — CHANGES_REQUESTED** [review]

File: lib/src/network/model/common/handle/handle.dart (hunk 2 above)
Comment (go-finn-nguyen) [reply lib/src/network/model/common/handle/handle.dart:13]: When an `MHandle` class is completed, it receives an `MResult` as a parameter. `MResult` already has 2 states `MResult.data `& `MResult.error` so we can handle it easily without rewriting them. You can see how it works in [state_widget.dart](https://github.com/GoldenOwlAsia/mobile-flutter-template/pull/3/commits/caadcf1ab43771ef699ce69a4622634b8b6b9efb#diff-932b450ab97ec0ce88bb2ff220da4bd576902ac4be6da94108ba66870e498b28) which I just update.

Comment (jimmy-to-goldenowl) [reply lib/src/network/model/common/handle/handle.dart:13]: I want to add more factory functions. Sometime we not create with MResult. so It is
```
const factory MHandle.initial() = _Initial;
const factory MHandle.loading() = _Loading;
const factory MHandle.success(T data) = _Success<T>;
const factory MHandle.errorr(e) = _Error;
const factory MHandle.result(MResult<T> result) = _Result<T>;
```

Comment (jimmy-to-goldenowl) [reply lib/src/network/model/common/handle/handle.dart:13]: Hi @go-finn-nguyen. I will merge this Pr. Please create another pr for this comment

**Pull request merged**